multicast/multicast.go

```diff
@@ -0,0 +1,435 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package multicast
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/netip"
+	"strings"
+	"sync"
+	"text/tabwriter"
+	"time"
+
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/node/addressing"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cilium/cilium-cli/defaults"
+	"github.com/cilium/cilium-cli/k8s"
+)
+
+const (
+	padding     = 3
+	minWidth    = 5
+	paddingChar = ' '
+)
+
+type Multicast struct {
+	client *k8s.Client
+	params Parameters
+}
+
+type Parameters struct {
+	CiliumNamespace  string
+	Writer           io.Writer
+	WaitDuration     time.Duration
+	MulticastGroupIP string
+	All              bool
+}
+
+func NewMulticast(client *k8s.Client, p Parameters) *Multicast {
+	return &Multicast{
+		client: client,
+		params: p,
+	}
+}
+
+func (m *Multicast) getCiliumNode(ctx context.Context, nodeName string) (v2.CiliumNode, error) {
+	ciliumNodes, err := m.client.ListCiliumNodes(ctx)
+	if err != nil {
+		return v2.CiliumNode{}, err
+	}
+	var ciliumNode v2.CiliumNode
+	for _, node := range ciliumNodes.Items {
+		if node.Name == nodeName {
+			ciliumNode = node
+		}
+	}
+	return ciliumNode, nil
+}
+
+func (m *Multicast) getCiliumInternalIP(nodeName string) (v2.NodeAddress, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), m.params.WaitDuration)
+	defer cancel()
+	ciliumNode, err := m.getCiliumNode(ctx, nodeName)
+	if err != nil {
+		return v2.NodeAddress{}, fmt.Errorf("unable to get cilium node: %w", err)
+	}
+	addrs := ciliumNode.Spec.Addresses
+	var ciliumInternalIP v2.NodeAddress
+	for _, addr := range addrs {
+		if addr.AddrType() == addressing.NodeCiliumInternalIP {
+			ip, err := netip.ParseAddr(addr.IP)
+			if err != nil {
+				continue
+			}
+			if ip.Is4() {
+				ciliumInternalIP = addr
+			}
+		}
+	}
+	if ciliumInternalIP.IP == "" {
+		return v2.NodeAddress{}, fmt.Errorf("ciliumInternalIP not found")
+	}
+	return ciliumInternalIP, nil
+}
+
+// ListGroup lists multicast groups in every node
+func (m *Multicast) ListGroups() error {
+	ctx, cancel := context.WithTimeout(context.Background(), m.params.WaitDuration)
+	defer cancel()
+
+	ciliumPodsList, err := m.client.ListPods(ctx, m.params.CiliumNamespace, metav1.ListOptions{LabelSelector: defaults.AgentPodSelector})
+	if err != nil {
+		return err
+	}
+	ciliumPods := ciliumPodsList.Items
+
+	var wg sync.WaitGroup
+	errCh := make(chan error, len(ciliumPods))
+	wg.Add(len(ciliumPods))
+
+	w := tabwriter.NewWriter(m.params.Writer, minWidth, 0, padding, paddingChar, 0)
```

Comment:
I have a common concern regarding all these commands. Output is in custom tabular format, which is not ideal as some users may want to consume it in other formats, like JSON, to automate some workflows. 

Could you consider separating data collection and data output in these methods? Fetch data from all the nodes, then output data based on the user's specified format.